plugin/storage/es/spanstore/writer.go

```diff
@@ -86,32 +99,28 @@ func spanIndexName(span *model.Span) string {
 
 // Check if index exists, and create index if it does not.
 func (s *SpanWriter) checkAndCreateIndex(indexName string, jsonSpan *jModel.Span) error {
-	// TODO: We don't need to check every write. Try to pull this out of WriteSpan.
-	exists, err := s.client.IndexExists(indexName).Do(s.ctx)
-	if err != nil {
-		return s.logError(jsonSpan, err, "Failed to find index", s.logger)
-	}
-	if !exists {
-		_, err = s.client.CreateIndex(indexName).Body(spanMapping).Do(s.ctx)
+	if !checkWriteCache(indexName, s.indexCache) {
+		_, err := s.client.CreateIndex(indexName).Body(spanMapping).Do(s.ctx)
```

Comment:
I can potentially see this failing if you have the case where the cache is empty so we do a createindex which fails. However, we've already added the index to the cache in the checkWriteCache above. All future writes will fail because the index was not created but the cache thinks it was. We need to only SET the index in the cache when CreateIndex succeeds.